### ISSUE TITLE
Failing Tests in `Test_OpenQasmDevice.cpp` when installing on MacOS

### ISSUE THREAD

**Issue by ystade** [PennyLaneAI/catalyst#1184]
### Expected behavior
When following the installation guide on https://docs.pennylane.ai/projects/catalyst/en/stable/dev/installation.html#minimal-building-from-source-guide for MacOS, after the step `make test` all tests should pass.

### Actual behavior
* After executing `make test` the last lines of the output are the following:
```sh
-------------------------------------------------------------------------------
Test PSWAP and ISWAP with BuilderType::Braket
-------------------------------------------------------------------------------
/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/tests/Test_OpenQasmDevice.cpp:529
...............................................................................

/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/tests/Test_OpenQasmDevice.cpp:529: FAILED:
due to unexpected exception with message:
  ModuleNotFoundError: No module named 'braket'

At:
    <string>(3): <module>

===============================================================================
test cases:  33 |  28 passed |  5 failed
assertions: 177 | 151 passed | 26 failed

make[1]: *** [test] Error 26
make: *** [test-runtime] Error 2
```

### System information
Machine: Mac M1, Mac M2, Mac M3, with MacOS  15.0.1 installed.
Python used in a virtualenv: 3.11.9 and 3.12.6

The output of `import pennylane as qml; qml.about()`:
```sh
Name: PennyLane
Version: 0.39.0.dev26
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author:
Author-email:
License: Apache License 2.0
Location: /Users/yannick/.pyenv/versions/catalyst/lib/python3.11/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, packaging, pennylane-lightning, requests, rustworkx, scipy, toml, typing-extensions
Required-by: amazon-braket-pennylane-plugin, PennyLane-Catalyst, PennyLane_Lightning, PennyLane_Lightning_Kokkos

Platform info:           macOS-15.0.1-arm64-arm-64bit
Python version:          3.11.9
Numpy version:           2.0.0
Scipy version:           1.13.0
Installed devices:
- default.clifford (PennyLane-0.39.0.dev26)
- default.gaussian (PennyLane-0.39.0.dev26)
- default.mixed (PennyLane-0.39.0.dev26)
- default.qubit (PennyLane-0.39.0.dev26)
- default.qutrit (PennyLane-0.39.0.dev26)
- default.qutrit.mixed (PennyLane-0.39.0.dev26)
- default.tensor (PennyLane-0.39.0.dev26)
- null.qubit (PennyLane-0.39.0.dev26)
- reference.qubit (PennyLane-0.39.0.dev26)
- lightning.qubit (PennyLane_Lightning-0.39.0.dev38)
- nvidia.custatevec (PennyLane-Catalyst-0.9.0.dev30)
- nvidia.cutensornet (PennyLane-Catalyst-0.9.0.dev30)
- oqc.cloud (PennyLane-Catalyst-0.9.0.dev30)
- softwareq.qpp (PennyLane-Catalyst-0.9.0.dev30)
- lightning.kokkos (PennyLane_Lightning_Kokkos-0.39.0.dev38)
- braket.aws.ahs (amazon-braket-pennylane-plugin-1.30.1)
- braket.aws.qubit (amazon-braket-pennylane-plugin-1.30.1)
- braket.local.ahs (amazon-braket-pennylane-plugin-1.30.1)
- braket.local.qubit (amazon-braket-pennylane-plugin-1.30.1)
```

### Additional information

- On Ubuntu this behavour could not be reproduced and everything has worked as expected.
- After installing the libraries into the system python version the test went through on one machine. Hence, the assumption is that for those tests the system python version is used instead of the version from the virtual environment.

**Comment by josh146** [comment]
Thanks for reporting this @ystade! This looks like it is because one of the tests depends on the braket plugin being installed (`pip install amazon-braket-pennylane-plugin`).

Ideally though we should ensure that this test is skipped if braket is not installed.

**Comment by dime10** [comment]
This is true, we normally skip pytests if an optional package is not installed. In this case it looks to be a C++ that was added not too long ago 🤔 

@ystade Normally all test and build dependencies are covered by `pip install -r requirements.txt`, did you run this line before building?

**Comment by burgholzer** [comment]
The problem here seems to be very macOS specific. Everything succeeds on a fresh install of Catalyst on Ubuntu.
It seems to be that these C++ tests which call Python via an embedded interpreter for some reason do not use the site-packages from the virtual environment that was used to build catalyst.
All the required dependencies (from the `requirements.txt`) are installed in that virtual environment. But they are not installed in the system python `site-packages`.
This is supported by the fact that manually installing `amazon-braked-sdk` and `numpy==2.0` in the global site packages fixes the errors in the tests, i.e., running, e.g.,
```
python3.12 -m pip install --user --break-system-packages amazon-braket-sdk "numpy==2.0"
```
where `python3.12` links to the exact version that was used to create the virtual environment (in which all dependencies, including catalyst, have been installed into as part of `make all`).
Naturally, that is not a recommended fix and the C++ tests should pick up the right Python `site-package` directory.
Note again, that Linux builds work out of the box and pick the packages from the virtual environment when running these tests.

**Comment by dime10** [comment]
Ah, thanks for surfacing the actual issue @burgholzer! We'll look into it.

**Comment by dime10** [comment]
Could this be because `python3` was not bound to the interpreter you expect? Our scripts by default will query `python3` in your environment and use that for everything.

**Comment by ystade** [comment]
I tried to reproduce the issue, however, now I run into another issue *before* the reported one but probably related.

### What I did so far from inside my local Catalyst repo?
```sh
git pull
rm -rf .venv
uv venv --python 3.12  # Python 3.13 throws an error in the CMake configuration
uv pip install -r requirements.txt
make all
```

Those steps above proceed successfully. However, when I execute the command `make test`, I get the following output:

<details><summary>Output of <code>make test</code></summary>
<code>
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C runtime test
Configure Catalyst Runtime
cmake -G Ninja -B /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build . \
                -DCMAKE_BUILD_TYPE=RelWithDebInfo \
                -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build/lib \
                -DCMAKE_C_COMPILER=/usr/bin/clang \
                -DMLIR_INCLUDE_DIRS=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//../mlir/llvm-project//mlir/include \
                -DCMAKE_CXX_COMPILER=/usr/bin/clang++ \
                -DCMAKE_C_COMPILER_LAUNCHER=/opt/homebrew/bin/ccache \
                -DCMAKE_CXX_COMPILER_LAUNCHER=/opt/homebrew/bin/ccache \
                -DENABLE_OPENQASM=ON \
                -DENABLE_OQD=OFF \
                -DENABLE_CODE_COVERAGE=OFF \
                -DPython_EXECUTABLE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/bin/python3 \
                -DENABLE_ADDRESS_SANITIZER=OFF \
                -DRUNTIME_ENABLE_WARNINGS=ON
-- Detected LLVM version - 3a8316216807d64a586b971f51695e23883331f7
-- ENABLE_OPENQASM is ON.
-- ENABLE_OQD is OFF.
-- ENABLE_CODE_COVERAGE is OFF.
-- ENABLE_ADDRESS_SANITIZER is OFF.
-- RUNTIME_ENABLE_WARNINGS is ON.
-- Found pybind11: /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/lib/python3.12/site-packages/pybind11/include (found version "2.13.6")
-- Configuring done (0.3s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build
cmake --build /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build --target runner_tests_qir_runtime runner_tests_mbqc_runtime runner_tests_openqasm -j8
[1/2] Linking CXX executable tests/runner_tests_qir_runtime
FAILED: tests/runner_tests_qir_runtime tests/runner_tests_qir_runtime-b12d07c_tests.cmake /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_qir_runtime-b12d07c_tests.cmake 
: && /usr/bin/clang++ -O2 -g -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  tests/CMakeFiles/runner_tests_qir_runtime.dir/Test_NullQubit.cpp.o -o tests/runner_tests_qir_runtime  -Wl,-rpath,/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/lib  _deps/catch2-build/src/libCatch2Main.a  lib/librtd_null_qubit.dylib  _deps/catch2-build/src/libCatch2.a  /Users/yannick/.local/share/uv/python/cpython-3.13.0-macos-aarch64-none/lib/libpython3.13.dylib  lib/librt_capi.dylib && cd /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests && /opt/homebrew/bin/cmake -D TEST_TARGET=runner_tests_qir_runtime -D TEST_EXECUTABLE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_qir_runtime -D TEST_EXECUTOR= -D TEST_WORKING_DIR=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests -D TEST_SPEC= -D TEST_EXTRA_ARGS= -D "TEST_PROPERTIES=SKIP_RETURN_CODE;4" -D TEST_PREFIX= -D TEST_SUFFIX= -D TEST_LIST=runner_tests_qir_runtime_TESTS -D TEST_REPORTER= -D TEST_OUTPUT_DIR= -D TEST_OUTPUT_PREFIX= -D TEST_OUTPUT_SUFFIX= -D TEST_DL_PATHS= -D TEST_DL_FRAMEWORK_PATHS= -D CTEST_FILE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_qir_runtime-b12d07c_tests.cmake -D ADD_TAGS_AS_LABELS=FALSE -P /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake
dyld[47679]: Library not loaded: /install/lib/libpython3.13.dylib
  Referenced from: <7C054B59-290B-3AD2-8D1B-80EEE0EDFD46> /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_qir_runtime
  Reason: tried: '/install/lib/libpython3.13.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/install/lib/libpython3.13.dylib' (no such file), '/install/lib/libpython3.13.dylib' (no such file)
CMake Error at /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake:82 (message):
  Error listing tests from executable
  '/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_qir_runtime':

    Result: Subprocess aborted
    Output: 

Call Stack (most recent call first):
  /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake:228 (catch_discover_tests_impl)


[2/2] Linking CXX executable tests/runner_tests_openqasm
FAILED: tests/runner_tests_openqasm tests/runner_tests_openqasm-b12d07c_tests.cmake /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_openqasm-b12d07c_tests.cmake 
: && /usr/bin/clang++ -O2 -g -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  tests/CMakeFiles/runner_tests_openqasm.dir/Test_OpenQasmBuilder.cpp.o tests/CMakeFiles/runner_tests_openqasm.dir/Test_OpenQasmDevice.cpp.o -o tests/runner_tests_openqasm  -Wl,-rpath,@loader_path -Wl,-rpath,@loader_path/../lib -Wl,-rpath,/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/lib  _deps/catch2-build/src/libCatch2Main.a  lib/librtd_openqasm.dylib  _deps/catch2-build/src/libCatch2.a  /Users/yannick/.local/share/uv/python/cpython-3.13.0-macos-aarch64-none/lib/libpython3.13.dylib  lib/librt_capi.dylib && cd /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests && /opt/homebrew/bin/cmake -D TEST_TARGET=runner_tests_openqasm -D TEST_EXECUTABLE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_openqasm -D TEST_EXECUTOR= -D TEST_WORKING_DIR=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests -D TEST_SPEC= -D TEST_EXTRA_ARGS= -D "TEST_PROPERTIES=SKIP_RETURN_CODE;4" -D TEST_PREFIX= -D TEST_SUFFIX= -D TEST_LIST=runner_tests_openqasm_TESTS -D TEST_REPORTER= -D TEST_OUTPUT_DIR= -D TEST_OUTPUT_PREFIX= -D TEST_OUTPUT_SUFFIX= -D TEST_DL_PATHS= -D TEST_DL_FRAMEWORK_PATHS= -D CTEST_FILE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_openqasm-b12d07c_tests.cmake -D ADD_TAGS_AS_LABELS=FALSE -P /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake
dyld[47678]: Library not loaded: /install/lib/libpython3.13.dylib
  Referenced from: <FA2EF1FB-6F24-30C4-A5EA-26903FF355EA> /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_openqasm
  Reason: tried: '/install/lib/libpython3.13.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/install/lib/libpython3.13.dylib' (no such file), '/install/lib/libpython3.13.dylib' (no such file)
CMake Error at /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake:82 (message):
  Error listing tests from executable
  '/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/tests/runner_tests_openqasm':

    Result: Subprocess aborted
    Output: 

Call Stack (most recent call first):
  /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_deps/catch2-src/extras/CatchAddTests.cmake:228 (catch_discover_tests_impl)


ninja: build stopped: subcommand failed.
make[1]: *** [test_runner] Error 1
make: *** [test-runtime] Error 2
</code>
</details>

**Comment by dime10** [comment]
Thanks for posting that @ystade! Yeah it seems CMake is trying to link against 3.13 libs for some reason, even though `Python_EXECUTABLE` is set correctly 🤔 Let me look into it

**Comment by dime10** [comment]
So the package itself doesn't rely on libpython (which allows you to embed the python interpreter into other executables), but the standalone runtime tests do. Chances are everything is fine if you were able to build the package from source, and you can also run all the integration tests without the runtime ones with `make pytest`.

Having said that, let's try to get to the bottom of this:
- can you make sure the python3.12 installation in your virtual environment also provides libpython? 
- do you have the output of `make all`? It could contain clues as to whether cmake was able to detect libpython (i.e. the Development.Embed component). If you don't have it anymore you can regenerate the relevant (runtime) part of it with `make clean-runtime` `make runtime`.

**Comment by ystade** [comment]
Ok, what I did now is the following:

1. I ran `make pytest`, which resulted in 1821 failed, 634 passed, 153 skipped, 160 xfailed, 6 xpassed, 24 warnings, 1 error. As far as I can see, the majority of (or all) failed tests fail because of the following error message: "symbol not found in flat namespace '_PyObject_DelAttr'"
2. As you requested, I made sure that libpython is available by running `uv pip install libpython` which installed the previously not installed package.
3. Unfortunately, I do not have the output of `make all` anymore but following comes the output of `make runtime` after I ran `make clean-runtime`

<details><summary>Output of <code>make runtime</code></summary>
<code>
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C runtime runtime ENABLE_OQD=OFF
Configure Catalyst Runtime
cmake -G Ninja -B /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build . \
                -DCMAKE_BUILD_TYPE=RelWithDebInfo \
                -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build/lib \
                -DCMAKE_C_COMPILER=/usr/bin/clang \
                -DMLIR_INCLUDE_DIRS=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//../mlir/llvm-project//mlir/include \
                -DCMAKE_CXX_COMPILER=/usr/bin/clang++ \
                -DCMAKE_C_COMPILER_LAUNCHER=/opt/homebrew/bin/ccache \
                -DCMAKE_CXX_COMPILER_LAUNCHER=/opt/homebrew/bin/ccache \
                -DENABLE_OPENQASM=ON \
                -DENABLE_OQD=OFF \
                -DENABLE_CODE_COVERAGE=OFF \
                -DPython_EXECUTABLE=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/bin/python3 \
                -DENABLE_ADDRESS_SANITIZER=OFF \
                -DRUNTIME_ENABLE_WARNINGS=ON
-- Detected LLVM version - 3a8316216807d64a586b971f51695e23883331f7
-- ENABLE_OPENQASM is ON.
-- ENABLE_OQD is OFF.
-- ENABLE_CODE_COVERAGE is OFF.
-- ENABLE_ADDRESS_SANITIZER is OFF.
-- RUNTIME_ENABLE_WARNINGS is ON.
-- Found pybind11: /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/lib/python3.12/site-packages/pybind11/include (found version "2.13.6")
-- Configuring done (0.3s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build
cmake --build /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime//build --target rt_capi rtd_null_qubit rtd_openqasm -j8
[6/13] Performing update step for 'lapacke-accelerate'
-- Fetching latest from the remote origin
[8/13] Performing configure step for 'lapacke-accelerate'
FAILED: _lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure 
cd /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-build && /opt/homebrew/bin/cmake -P /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure-RelWithDebInfo.cmake && /opt/homebrew/bin/cmake -E touch /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure
CMake Error at /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure-RelWithDebInfo.cmake:49 (message):
  Command failed: 1

   '/opt/homebrew/bin/cmake' '--preset accelerate-lapacke32' '-DCMAKE_INSTALL_PREFIX=/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/install' '-GNinja' '-S' '/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate' '-B' '/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-build'

  See also

    /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/runtime/build/_lapacke-accelerate/src/lapacke-accelerate-stamp/lapacke-accelerate-configure-*.log


ninja: build stopped: subcommand failed.
make[1]: *** [runtime] Error 1
make: *** [runtime] Error 2
</code>
</details>

The command `make all` (that previously executed successfully) results now in the same error. Taking a look in the logs referred to in the error message yields the following:

<details><summary><code>lapacke-accelerate-configure-out.log</code></summary>
<code>
-- The Darwin kernel version is: 24.4.0
-- The MacOS version is: >=15.0
-- The BLAS/LAPACK version provided by the Accelerate framework is: 3.11.0
-- The MacOS SDK is: 
-- The minimum compatible MacOS SDK version is: 15.0
-- Configuring incomplete, errors occurred!
</code>
</details>

<details><summary><code>lapacke-accelerate-configure-err.log</code></summary>
<code>
CMake Error at CMakeLists.txt:75 (message):
  Please install a more recent XCode for a compatible MacOS SDK.
</code>
</details>

Between my last message and this message, I updated my OS. To avoid any problems with some artefacts from the older OS, I deleted all build files to start basically from a fresh copy of Catalyst. I also executed `git pull` to get the latest updates. Still, I get the same error message from above without any change.

**Comment by ystade** [comment]
Btw, the command `make STRICT_WARNINGS=OFF mlir` executes successfully but only with `STRICT_WARNINGS=OFF` otherwise the process ends with the following error message:
<details><summary>End of output of <code>make mlir</code></summary>
<code>
cmake --build /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir//build --target check-dialects quantum-lsp-server catalyst-cli check-unit-tests
[3/135] Building CXX object python/CMakeFiles/QuantumPythonModules.extension._mlirLinalgPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp.o
FAILED: python/CMakeFiles/QuantumPythonModules.extension._mlirLinalgPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp.o 
/opt/homebrew/bin/ccache /usr/bin/clang++ -DMLIR_PYTHON_PACKAGE_PREFIX=mlir_quantum. -DQuantumPythonModules_extension__mlirLinalgPasses_dso_EXPORTS -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/include -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/llvm/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/tools/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/stablehlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build/stablehlo -isystem /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/lib/python3.12/site-packages/pybind11/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=gnu++20 -arch arm64 -fPIC -fvisibility=hidden   -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -UNDEBUG -frtti -fexceptions -flto -Werror -MD -MT python/CMakeFiles/QuantumPythonModules.extension._mlirLinalgPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp.o -MF python/CMakeFiles/QuantumPythonModules.extension._mlirLinalgPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp.o.d -o python/CMakeFiles/QuantumPythonModules.extension._mlirLinalgPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp.o -c /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp
In file included from /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/LinalgPasses.cpp:11:
In file included from /usr/local/include/pybind11/pybind11.h:12:
In file included from /usr/local/include/pybind11/detail/class.h:12:
In file included from /usr/local/include/pybind11/attr.h:14:
In file included from /usr/local/include/pybind11/cast.h:15:
/usr/local/include/pybind11/detail/type_caster_base.h:568:13: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]
  568 |             default:
      |             ^
1 error generated.
[4/135] Building CXX object python/CMakeFiles/QuantumPythonModules.extension._mlirAsyncPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp.o
FAILED: python/CMakeFiles/QuantumPythonModules.extension._mlirAsyncPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp.o 
/opt/homebrew/bin/ccache /usr/bin/clang++ -DMLIR_PYTHON_PACKAGE_PREFIX=mlir_quantum. -DQuantumPythonModules_extension__mlirAsyncPasses_dso_EXPORTS -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/include -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/llvm/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/tools/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/stablehlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build/stablehlo -isystem /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/lib/python3.12/site-packages/pybind11/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=gnu++20 -arch arm64 -fPIC -fvisibility=hidden   -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -UNDEBUG -frtti -fexceptions -flto -Werror -MD -MT python/CMakeFiles/QuantumPythonModules.extension._mlirAsyncPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp.o -MF python/CMakeFiles/QuantumPythonModules.extension._mlirAsyncPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp.o.d -o python/CMakeFiles/QuantumPythonModules.extension._mlirAsyncPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp.o -c /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp
In file included from /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/AsyncPasses.cpp:12:
In file included from /usr/local/include/pybind11/pybind11.h:12:
In file included from /usr/local/include/pybind11/detail/class.h:12:
In file included from /usr/local/include/pybind11/attr.h:14:
In file included from /usr/local/include/pybind11/cast.h:15:
/usr/local/include/pybind11/detail/type_caster_base.h:568:13: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]
  568 |             default:
      |             ^
1 error generated.
[5/135] Building CXX object python/CMakeFiles/QuantumPythonModules.extension._mlirGPUPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp.o
FAILED: python/CMakeFiles/QuantumPythonModules.extension._mlirGPUPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp.o 
/opt/homebrew/bin/ccache /usr/bin/clang++ -DMLIR_PYTHON_PACKAGE_PREFIX=mlir_quantum. -DQuantumPythonModules_extension__mlirGPUPasses_dso_EXPORTS -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/include -I/Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/llvm/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/build/tools/mlir/include -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/stablehlo -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/mlir-hlo/bazel-build/stablehlo -isystem /Library/Frameworks/Python.framework/Versions/3.12/include/python3.12 -isystem /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/.venv/lib/python3.12/site-packages/pybind11/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -O3 -DNDEBUG -std=gnu++20 -arch arm64 -fPIC -fvisibility=hidden   -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -UNDEBUG -frtti -fexceptions -flto -Werror -MD -MT python/CMakeFiles/QuantumPythonModules.extension._mlirGPUPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp.o -MF python/CMakeFiles/QuantumPythonModules.extension._mlirGPUPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp.o.d -o python/CMakeFiles/QuantumPythonModules.extension._mlirGPUPasses.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp.o -c /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp
In file included from /Users/yannick/Documents/TUM/CDA/Research/playgrounds/Catalyst/catalyst/mlir/llvm-project/mlir/lib/Bindings/Python/GPUPasses.cpp:12:
In file included from /usr/local/include/pybind11/pybind11.h:12:
In file included from /usr/local/include/pybind11/detail/class.h:12:
In file included from /usr/local/include/pybind11/attr.h:14:
In file included from /usr/local/include/pybind11/cast.h:15:
/usr/local/include/pybind11/detail/type_caster_base.h:568:13: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]
  568 |             default:
      |             ^
1 error generated.
[12/135] Building CXX object python/CMakeFiles/QuantumPythonModules.extension._mlirDialectsSparseTensor.dso.dir/__/llvm-project/mlir/lib/Bindings/Python/DialectSparseTensor.cpp.o
ninja: build stopped: subcommand failed.
make[1]: *** [dialects] Error 1
make: *** [mlir] Error 2
</code>
</details>

**Comment by dime10** [comment]
> Between my last message and this message, I updated my OS. To avoid any problems with some artefacts from the older OS, I deleted all build files to start basically from a fresh copy of Catalyst. I also executed git pull to get the latest updates. Still, I get the same error message from above without any change.

Ah yes, unfortunately on macOS if you update the system (specifically the compiler/sdk) existing cmake build files will be invalidated (I have the same problem). You can still use already built binaries, but not any build artifacts that would rely on any generated cmake or build system caches and files. So clearing the build dir is the right move here.

**Comment by dime10** [comment]
> Btw, the command make STRICT_WARNINGS=OFF mlir executes successfully but only with STRICT_WARNINGS=OFF otherwise the process ends with the following error message:

Thanks for reporting, looks like the pybind11 headers do not build warning-free in your configuration. I'll have to update the build script to ensure pybind11 headers are treated as system headers (and thus do not raise warnings)!

---

So with these separate issues (OS update, header warnings) out of the way, what is your status with respect to the original issue? Did installing `libpython` make the failure go away that was using different versions of Python?

**Comment by ystade** [comment]
I, unfortunately, do not reach the point where I could check whether the initial issue still exists or whether installing `libpython` solved the issue. Meanwhile to get rid of literally all build files, I completely removed the Catalyst repository and cloned a fresh copy of it. Still, I get the same error with lapacke as reported above.

I tried to solve the issue and performed the following steps:
1. since cmake v4.0 `CMAKE_OSX_SYSROOT` defaults to empty which results in errors in the lapacke configuration. I fixed that issue by adding `-DCMAKE_OSX_SYSROOT=$$(xcrun --sdk macosx --show-sdk-path)` in line 85 in `runtime/Makefile` (this will lead to issues on other platforms, I assume).
2. Afterward, I run into the next issue "Compatibility with CMake < 3.5 has been removed from CMake" occuring in the configuration of lapacke itself. I went into `runtime/build/_lapacke-accelerate/src/lapacke-accelerate-build/_deps/reference-lapack-src/` and performed `git switch master` (A change that will be lost after the next clean). Then, I run `make runtime` from the root directory in the catalyst repository again. Then the configuration of lapacke run through successfully.
3. Now, an error occured in the build process of lapacke. The linker could not find a couple of symbols. I'll include the last 20 lines of the output below.

<details><summary>Last couple lines of Linker error</summary>
<code>
  "_zunmrq_64_", referenced from:
      _LAPACKE_zunmrq_work_64 in lapacke_zunmrq_work.c.o
      _LAPACKE_zunmrq_work_64 in lapacke_zunmrq_work.c.o
  "_zunmrz_64_", referenced from:
      _LAPACKE_zunmrz_work_64 in lapacke_zunmrz_work.c.o
      _LAPACKE_zunmrz_work_64 in lapacke_zunmrz_work.c.o
  "_zunmtr_64_", referenced from:
      _LAPACKE_zunmtr_work_64 in lapacke_zunmtr_work.c.o
      _LAPACKE_zunmtr_work_64 in lapacke_zunmtr_work.c.o
  "_zupgtr_64_", referenced from:
      _LAPACKE_zupgtr_work_64 in lapacke_zupgtr_work.c.o
      _LAPACKE_zupgtr_work_64 in lapacke_zupgtr_work.c.o
  "_zupmtr_64_", referenced from:
      _LAPACKE_zupmtr_work_64 in lapacke_zupmtr_work.c.o
      _LAPACKE_zupmtr_work_64 in lapacke_zupmtr_work.c.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
</code>
</details>

Right now, I do not have the capacity to dig deeper into that issue with lapacke, sorry.

**Comment by burgholzer** [comment]
Just a side comment from an innocent bystander:
These issues with the latest CMake major release will probably plague the ecosystem for a while because people hardly tend to update their build systems and frequently ignore warnings.
For this particular one, you should be able to work around the issues by explicitly overriding the minimum version via `-DCMAKE_POLICY_VERSION_MINIMUM=3.5` (or higher). See https://cmake.org/cmake/help/latest/variable/CMAKE_POLICY_VERSION_MINIMUM.html for more details.

**Comment by ystade** [comment]
Thank you @burgholzer for the comment. Setting the variable as proposed in the place where cmake is invoked, i.e., in line 85 of `runtime/Makefile` (where I added the definition of sysroot before) did not solve the issue as the value of the variable is not properly propagated into the configure process of lapacke.

I tried something else, that worked in the end.
1. I invoked `make clean-runtime` to remove build files especially the entire clone of lapacke. (I did not switch to the master branch as in last message).
2. I performed `export CMAKE_POLICY_VERSION_MINIMUM=3.5` to set the respective variable for all cmake configuration processes in this environment.
3. Now, `make runtime` executed successfully. I could also execute `make all` without error.

**Comment by burgholzer** [comment]
Yeah, setting it as an environment variable is the preferred way for categorically overriding the setting for a whole build.
Glad to hear that it helped to get your build working.

An alternative would have been to use an older CMake Version. CMake has a PyPI distribution, so you could have probably also just done `uv tool install "cmake<4"` and use that.

**Comment by dime10** [comment]
> In the next step, I run `make test`. This gave the following result
> 
> ===============================================================================
> test cases:  34 |  28 passed |  6 failed
> assertions: 183 | 151 passed | 32 failed
> All failed tests failed because `ModuleNotFoundError: No module named 'numpy'` even though numpy is indeed installed.

Thanks @ystade, I could reproduce this issue locally as well. Apparently this is somewhat of a tricky issue, when spawning a python interpreter from C++ together with a virtual environment, not sure why this has never come up before 🤔 

What happens is that the interpreter spawned through from C++ (using libpython.dylib) is using the global environment, and it can't find the `numpy` package there. We only do this is in C++ tests, as Python tests already have a running interpreter instance, and we also only need this for our OpenQASM component.

We'll try to find a fix for our config scripts, but in the meantime you can simply disable the OpenQASM component `export ENABLE_OPENQASM=OFF`, avoid running the runtime tests (e.g. `make pytest` instead of `make test`), or install the openqasm test requirements in the global environment (`deactivate && pip install numpy amazon-braket-sdk`). Sorry for the incovenience!

**Comment by dime10** [comment]
> As you requested, I made sure that libpython is available by running `uv pip install libpython` which installed the previously not installed package.

@ystade I didn't think libpython would be a Python package, it would be part of a python installation or system package (some Linux images for example only install the headers but not the shared library for the interpreter). Is this (https://pypi.org/project/libpython/) what you installed? If so I would recommend removing it.
